## Abstract {.page_break_before}

The assignment of genomes and taxonomy to metagenome data underlies
many microbiome studies. Here we describe two algorithms for
compositional analysis of metagenome sequencing data. We first develop
a sketching technique, _Scaled MinHash_, that
supports Jaccard containment estimation.
We implement _Scaled MinHash_ in the sourmash software and demonstrate
large-scale containment searches of metagenomes using all 700,000 currently
available microbial reference genomes.
We next frame shotgun
metagenome compositional analysis in terms of min-set-cover, i.e. as
the problem of finding a minimum collection of reference genomes
that "cover" the known portion of a metagenome. We implement a greedy
approximate solution that uses  _Scaled MinHash_ sketches.
Finally, we show that by
linking genomes to their taxonomic lineages, we can provide a
lightweight and precise method for taxonomic classification of
metagenome content.  sourmash is available as open source under the
BSD 3-Clause license at github.com/dib-lab/sourmash/.
