## Abstract {.page_break_before}

The accurate assignment of genomes and taxonomy to metagenome data is
a significant challenge in microbiome studies. Here we describe
_Scaled MinHash_, an extension of MinHash k-mer sketching that permits
rapid and accurate compositional analysis of shotgun metagenome data
sets.  We implement this approach in the sourmash software, in order to
support large-scale Jaccard containment searches across all 700,000
currently available microbial reference genomes.  We then approach
shotgun metagenome compositional analysis as a min-set-cover problem,
i.e. as a problem of finding the minimal collection of reference
genomes for a metagenome. We implement a greedy approximate solution
using _Scaled MinHash_. Finally, we show that by linking genomes to
their taxonomy, we can provide a lightweight and precise method for
taxonomic classification of metagenome content.  sourmash is available
as open source under the BSD 3-Clause license at
github.com/dib-lab/sourmash/.
