## Abstract {.page_break_before}

The assignment of reference genomes and taxonomy to metagenome data underlies
many microbiome studies. Here we describe two algorithms for
compositional analysis of metagenome sequencing data. We first investigate
the _FracMinHash_ sketching technique, a derivative of modulo hash that
supports Jaccard containment estimation between sets of different size.
We implement _FracMinHash_ in the sourmash software, evaluate its
accuracy, and demonstrate
large-scale containment searches of metagenomes using all 700,000 currently
available microbial reference genomes.
We next frame shotgun
metagenome compositional analysis as the problem of finding a minimum
collection of reference genomes that "cover" the known k-mers in a metagenome,
a minimum set cover problem.
We implement a greedy
approximate solution using  _FracMinHash_ sketches, and evaluate
its accuracy in taxonomic assignment using a CAMI community benchmark.
Finally, we show that the minimum metagenome cover can be used to guide the
select of reference genomes for read mapping.
sourmash is available as open source under the
BSD 3-Clause license at github.com/dib-lab/sourmash/.
