## Abstract {.page_break_before}

The interpretation of metagenome data
The assignment of genomes and taxonomy to metagenome data underlies
many microbiome studies. Here we describe two algorithms for
compositional analysis of metagenome sequencing data. We first develop
a sketching technique, _Scaled MinHash_, that
supports Jaccard containment estimation between sets of different size.
We implement _Scaled MinHash_ in the sourmash software and demonstrate
large-scale containment searches of metagenomes using all 700,000 currently
available microbial reference genomes.
We next frame shotgun
metagenome compositional analysis in terms of min-set-cover, i.e. as
the problem of finding a minimum collection of reference genomes
that "cover" the known k-mers in a metagenome. We implement a greedy
approximate solution using  _Scaled MinHash_ sketches, and evaluate
its accuracy in taxonomic assignment using a CAMI community benchmark.
Finally, we show that the minimum set cover can be used for accurate read
mapping.
sourmash is available as open source under the
BSD 3-Clause license at github.com/dib-lab/sourmash/.
