## Abstract {.page_break_before}

The accurate assignment of genomes and taxonomy to microbiome samples
is a significant challenge in microbiome studies. Here we describe
_Scaled MinHash_, an extension of MinHash sketching that permits rapid
and accurate compositional analysis of shotgun metagenome data sets
using k-mers.  We implement this approach in the sourmash software,
together with database indices that support
large-scale Jaccard containment searches across all
available reference genomes. We then approach shotgun
metagenome compositional analysis as a min-set-cover challenge and
implement a greedy approximate solution using _Scaled
MinHash_. We show that this solution finds the minimal collection of
genomes needed to serve as a reference for all mappable metagenome
reads, and extend it to provide a lightweight and precise method
for taxonomic classification of metagenome content.  sourmash is
freely available under the BSD 3-Clause license at
github.com/dib-lab/sourmash/.
