## Abstract {.page_break_before}

The accurate assignment of genomes and taxonomy to microbiome samples
is a significant challenge in microbiome studies. Here
we describe _Scaled MinHash_, an extension of MinHash sketching that
permits rapid and accurate compositional analysis of shotgun
metagenome data sets using k-mers.  We implement this approach in the
sourmash software, and provide database indexing implementations that
support large-scale Jaccard similarity and containment searches across
all available reference genomes. We next frame the problem of shotgun
metagenome compositional analysis as a min-set-cover challenge and
implement a greedy approximate solution in sourmash using _Scaled
MinHash_. We show that this approach finds the minimal collection of
reference genomes necessary to map all possible metagenome reads, and
provides a lightweight and precise method for taxonomic classification
of metagenome content.  sourmash is freely available under the BSD
3-Clause license at github.com/dib-lab/sourmash/.
