## Abstract {.page_break_before}

The assignment of genomes and taxonomy to metagenome data is an
important technique underlying many microbiome studies. Here we
describe a k-mer sketching technique, _Scaled MinHash_, that is an
extension of MinHash. _Scaled MinHash_ permits rapid and accurate
compositional analysis of shotgun metagenome data sets.  We implement
this approach in the sourmash software, which supports large-scale
Jaccard containment searches across all 700,000 currently available
microbial reference genomes.  We then frame shotgun metagenome
compositional analysis in terms of min-set-cover, i.e. as the problem
of finding the minimal collection of reference genomes for a
metagenome. We implement a greedy approximate solution using _Scaled
MinHash_. Finally, we show that by linking genomes to their taxonomic
lineages, we can provide a lightweight and precise method for
taxonomic classification of metagenome content.  sourmash is available
as open source under the BSD 3-Clause license at
github.com/dib-lab/sourmash/.
