# Introduction

Shotgun metagenomics measures the sequence content of microbial communities.

Compositional analysis of shotgun metagenome samples addresses the
question of what reference genomes should be used for functional
and taxononomic interpretation of metagenome content.

Below, we describe a lightweight approach to compositional analysis of
shotgun metagenome samples using k-mer hashing.

We first define _Scaled MinHash_, an extension of MinHash sketching
that supports lightweight containment estimation for metagenome
datasets.  We implement _Scaled MinHash_ in
Python and Rust, and show that it is competitive in accuracy with
other containment estimation approaches.  We further provide efficient
indexed database implementations that support similarity and containment
search across hundreds of thousands of reference genomes.

We next frame reference-based metagenome content analysis as a
min-set-cov problem, in which we seek the _minimum_ number of genomes
in the reference database necessary to cover the known content of a
metagenome.  We implement the known best-polynomial-time greedy
approximate to the min-set-cov problem using _Scaled MinHash_, and
show that it is recovers a minimum set of reference genomes for the mappable
reads in a metagenome.

Finally, we implement a simple taxonomic classification approach on
top of min-set-cov, in which we transfer the taxonomy of the genomes
from the set cover to the metagenome. We show that this permits
precise and lightweight classification of metagenome content across
all taxonomic levels.
