# Introduction

Shotgun metagenomics samples the DNA sequence content of microbial
communities.

Compositional analysis of shotgun metagenome samples has the goal of
identifying what functions and taxonomic units are present in the
data.  Function and taxonomy is typically inferred from available
reference genomes and gene catalogs, via direct genomic alignment
(cite biobakery, MEGAN-LR), gene search (cite mmseqs, eggnog etc.), or k-mer matches
(Kraken).  For many of these methods, the substantial increase in the
number of available reference genomes (1.1m in GenBank as of XYZ)
presents a significant practical obstacle to comprehensive
compositional analyses, and most methods choose representative subsets
of available genomic information to analyze.

Here, we describe a lightweight and scalable approach to compositional
analysis of shotgun metagenome data. Our approach tackles the
selection of appropriate reference genomes for downstream analysis and
provides a computationally efficient method for taxonomic
classification of metagenome data.  Our implementation in the
`sourmash` software can make use of all currently available microbial
genomes.

We first define _Scaled MinHash_, an extension of MinHash sketching
that supports containment estimation for metagenome
datasets using k-mers.  We implement _Scaled MinHash_ in a Python and
Rust package, `sourmash`, and show that it is competitive in
accuracy with other containment estimation approaches.

We next frame reference-based metagenome content analysis as a
min-set-cov problem, where we determine the _minimum_ number of
genomes from a reference database needed to cover the identifiable
genomic content from a metagenome.  We implement a
best-polynomial-time greedy approximation to the min-set-cov problem
using _Scaled MinHash_ in `sourmash`. This technique provides an
iterative decomposition of metagenomes into genome matches.

To evaluate the accuracy of our min-set-cov procedure, we implement a
simple taxonomic classification approach in which we use the taxonomy
of the genomes from the set cover to define the taxonomy of the
metagenome content. We show that this permits precise and lightweight
classification of metagenome content across all taxonomic levels.

Finally, we show that the minimum set covers for several metagenomes
contain only a small subset of reference genomes even when using very
large and redundant databases, and demonstrate that this subset can be
used to map the metagenome reads in concordance with the estimates
from _Scaled MinHash_. Thus, _Scaled MinHash_ combined with
min-set-cov provides a lightweight, accurate, and scalable way to
estimate the composition of metagenomes using a large reference
database.
