# Introduction

Shotgun DNA sequencing of microbial communities is an important
technique for studying host-associated and environmental microbiomes
[@doi:10.1038/s41587-020-0718-6; @doi:10.1111/1462-2920.15314].  By
sampling the genomic content of microbial communities, shotgun
metagenomics enables the taxonomic and functional characterization of
microbiomes
[@doi:10.1038/s41586-019-1238-8; @doi:10.1126/science.1261359].
However, this characterization relies critically on the methods and
databases used to interpret the sequencing data
[@doi:10.1038/d41586-019-01654-0; @doi:10.1038/s41592-021-01141-3; @doi:10.1101/2021.07.12.451567; @doi:10.1093/bib/bbx120].

Metagenome function and taxonomy are typically inferred from available
reference genomes and gene catalogs, via direct genomic alignment
[@biobakery3;@meganlr], large-scale protein search [@mmseqs2;@motu2;@eggnog],
or k-mer matches [@kraken2;@kaiju].  For many of these methods, the
substantial increase in the number of available microbial reference
genomes (1.1m in GenBank as of November 2021) presents a significant
practical obstacle to comprehensive compositional analyses. Most
methods choose representative subsets of available genomic information
to analyze; for example, bioBakery 3 provides a database containing
99.2k reference genomes [@biobakery3].  Scaling metagenome analysis
approaches to make use of the rapidly increasing size of GenBank is an
active endeavor in the field [@ganon; @metalign].

Here, we describe a lightweight and scalable approach to compositional
analysis of shotgun metagenome data based on finding the minimum set
of reference genomes that accounts for all known k-mers in a
metagenome - a "minimum metagenome cover".  We use a mod-hash based
sketching approach for k-mers to reduce memory requirements
[@broder_minhash], and implement a polynomial-time greedy
approximation algorithm for the minimum set cover analysis
[@polynomial_minsetcov].

Our approach tackles the selection of appropriate reference genomes
for downstream analysis and provides a computationally efficient
method for taxonomic classification of metagenome data.  Our
implementation in the `sourmash` open source software package works with
reference databases containing a million or more microbial genomes and
supports multiple taxonomies and private databases.
