# Introduction

Shotgun metagenomics samples the DNA sequence content of microbial communities.

Compositional analysis of shotgun metagenome samples has the goal
of identifying what reference genomes to use for functional
and taxononomic interpretation of metagenome content.

The substantial increase in the number of available reference genomes
presents a significant practical obstacle to comprehensive
compositional analyses.

Here, we describe a lightweight approach to compositional analysis of
shotgun metagenome data. Our approach tackles the selection
of appropriate reference genomes and provides a lightweight method for
taxonomic classification of metagenome data.

We first define _Scaled MinHash_, an extension of MinHash sketching
that supports lightweight containment estimation for metagenome
datasets using k-mers.  We implement _Scaled MinHash_ in Python and
Rust, and show that it is competitive in accuracy with other
containment estimation approaches.

We next frame reference-based metagenome content analysis as a
min-set-cov problem, in which we seek the _minimum_ number of genomes
from the reference database necessary to cover the identifiable
genomic content of a metagenome.  We implement a best-polynomial-time
greedy approximation to the min-set-cov problem using _Scaled
MinHash_, and show that it recovers a minimum set of reference genomes
for the mappable reads in a metagenome.

Finally, we implement a simple taxonomic classification approach on
top of min-set-cov, in which we transfer the taxonomy of the genomes
from the set cover to the metagenome. We show that this permits
precise and lightweight classification of metagenome content across
all taxonomic levels.
