# Introduction

Shotgun DNA sequencing of microbial communities is an important
technique for studying host-associated and environmental microbiomes 
[@doi:10.1038/s41587-020-0718-6; @doi:10.1111/1462-2920.15314].
By sampling the DNA sequence content of microbial communities, shotgun
metagenomics enables the taxonomic and functional characterization of
microbiomes [@doi:10.1038/s41586-019-1238-8; @doi:10.1126/science.1261359]. 
However, this characterization relies critically on the methods and 
databases used to interpret the sequencing data 
[@doi:10.1038/d41586-019-01654-0; @doi:10.1038/s41592-021-01141-3; 
@doi:10.1101/2021.07.12.451567; @doi:10.1093/bib/bbx120].

Metagenome function and taxonomy is typically inferred from available
reference genomes and gene catalogs, via direct genomic alignment
[@doi:10.7554/eLife.65088;@doi:10.1186/s13062-018-0208-7], large-scale
protein search
[@doi:10.1093/bioinformatics/btab184;@doi:10.1093/nar/gky1085], or
k-mer matches
[@doi:10.1186/s13059-019-1891-0;@doi:10.1038/ncomms11257].  For many
of these methods, the substantial increase in the number of available
reference genomes (1.1m in GenBank as of DATE) presents a significant
practical obstacle to comprehensive compositional analyses, and most
methods choose representative subsets of available genomic information
to analyze; for example, bioBakery 3 provides a database containing
99.2k reference genomes [@doi:10.7554/eLife.65088].

Here, we describe a lightweight and scalable approach to compositional
analysis of shotgun metagenome data based on finding the minimum set
of reference genomes that accounts for all known k-mers in a
metagenome.  We use a mod-hash based sketching approach for k-mers to
reduce memory requirements, and implement a polynomial-time greedy
approximation algorithm for the minimum set analysis.

Our approach tackles the selection of appropriate reference genomes
for downstream analysis and provides a computationally efficient
method for taxonomic classification of metagenome data.  Our
implementation in the open source `sourmash` software works with
reference databases containing a million or more microbial genomes and
supports multiple taxonomies and private databases.

<!--

We first describe _FracMinHash_, a sketching technique based on
mod-hash (cite Broder), 
that supports containment estimation for metagenome
datasets using k-mers.  We implement _FracMinHash_ in a Python and
Rust package, `sourmash`, and demonstrate that it is competitive in
accuracy with other containment estimation approaches.

We next frame reference-based metagenome content analysis as a
min-set-cov problem, where we determine the _minimum_ number of
genomes from a reference database needed to cover the identifiable
genomic content from a metagenome.  We implement a
best-polynomial-time greedy approximation to the min-set-cov problem
using _FracMinHash_ in `sourmash`. This technique provides an
iterative decomposition of metagenomes into genome matches.

To evaluate the accuracy of our min-set-cov procedure, we implement a
simple taxonomic classification approach in which we use the taxonomy
of the genomes in the set cover to define the taxonomy of the
metagenome content. We show that this permits precise and lightweight
classification of metagenome content across all taxonomic levels.

Finally, we show that the minimum set covers for several metagenomes
contain only a small subset of reference genomes even when using very
large and redundant databases, and demonstrate that this subset can be
used to map the metagenome reads in concordance with the estimates
from _FracMinHash_. Thus, _FracMinHash_ combined with
min-set-cov provides a lightweight, accurate, and scalable way to
estimate the composition of metagenomes using a large reference
database.
-->
