# Methods

## Analytical analysis of FracMinHash

Given two arbitrary sets $A$ and $B$ which are subsets of a domain
$\Omega$, the containment index $C(A,B)$ is defined as
$C(A,B):=\frac{\vert A \cap B \vert}{\vert A \vert}$. Let $h$ be a
perfect hash function $h~:~\Omega \rightarrow~[0,H]$ for some $H\in
\mathbb{R}$. For a *scale factor* $s$ where $0 \le s \le 1$, a
FracMinHash sketch of a set $A$ is defined as follows:

{% raw %}
```{=latex}
\begin{equation}
    %\mathbf{FRAC}_S(A) = \left\{\,h(a) \mid \forall a \in A\ {\rm s.t.}\ h(a) \leq Hs\right\}.
    \mathbf{FRAC}_S(A) = \left\{\,h(a) \mid \forall a \in A\ {\rm s.t.}\ h(a) \leq Hs\right\}.
\end{equation}
```
{% endraw %}

The scale factor $s$ is a tunable parameter that can modify the size of the sketch. Using this FracMinHash sketch, we define the FracMinHash estimate of the containment index $\hat{C}_\text{frac}(A,B)$ as follows:

{% raw %}
```{=latex}
\begin{equation}
    \hat{C}_\text{frac}(A,B):=\frac{\vert \mathbf{FRAC}_S(A) \cap \mathbf{FRAC}_S(B)\vert }{\vert \mathbf{FRAC}_S(A)\vert}.
\end{equation}
```
{% endraw %}

For notational simplicity, we define $X_A := \vert \mathbf{FRAC}_S(A)
\vert$. Observe that if one views $h$ as a uniformly distributed
random variable, we have that $X_A$ is distributed as a binomial
random variable: $X_A \sim {\rm Binom}(|A|, s)$. Furthermore, if
$A\cap B \neq \emptyset$ where both $A$ and $B$ are non-empty sets, then
$X_A$ and $X_B$ are independent when the probability of success is
strictly smaller than $1$. Using these notations, we compute the
expectation of $\hat{C}_\text{frac}(A,B)$.

<!-- CTB: where does comment ,\hat{s}, come in? -->

**Theorem 1:**
For $0<s<1$, if $A$ and $B$ are two distinct sets such that $A \cap B$ is non-empty,
{% raw %}
```{=latex}
\begin{align}
\mathrm{E}\left[\hat{C}_\text{frac}(A,B) \unicode{x1D7D9}_{\vert \mathbf{FRAC}_S(A) \vert>0} \right] =
\frac{\vert A\cap B \vert}{\vert A \vert} \left(1-(1-s)^{\vert A\vert}\right).
\end{align}
```
{% endraw %}


*Proof.* Using the notation introduced previously, observe that 

{% raw %}
```{=latex}
\begin{align}
\hat{C}_\text{frac}(A,B) \unicode{x1D7D9}_{\vert \mathbf{FRAC}_S(A) \vert>0} = \frac{X_{A\cap B}}{X_{A\cap B} + X_{A\setminus B}} \unicode{x1D7D9}_{X_{A\cap B} + X_{A\setminus B}>0},
\end{align}
```
{% endraw %}

and that the random variables $X_{A\cap B}$ and $X_{A\setminus B}$ are independent (which follows directly from the fact that $A \cap B$ is non-empty, and because $A$ and $B$ are distinct, $A \setminus B$ is also non-empty).
We will use the following fact from standard calculus:

{% raw %}
```{=latex}
\begin{align}
    \int_0^1 x t^{x+y-1}\, dt = \frac{x}{x+y} \unicode{x1D7D9}_{x+y>0}.
\end{align}
```
{% endraw %}

Then using the moment generating function of the binomial distribution, we have

{% raw %}
```{=latex}
\begin{align}
    \mathrm{E}\left[t^X_{A\cap B}\right] &= (1-s+st)^{\vert A \cap B \vert}\\
    \mathrm{E}\left[t^X_{A\setminus B}\right] &= (1-s+st)^{\vert A \setminus B \vert}.
\end{align}
```
{% endraw %}

We also know by continuity that 
{% raw %}
```{=latex}
\begin{align}
    \mathrm{E}\left[X_{A\cap B} \, t^{X_{A\cap B}-1}\right] &= \frac{d}{dt} (1-s+st)^{\vert A \cap B \vert}\\
    &= \vert A\cap B \vert s (1-s+st)^{\vert A\cap B\vert-1}.
\end{align}
```
{% endraw %}

Using these observations, we can then finally calculate that 


{% raw %}
```{=latex}
\begin{align}
    \mathrm{E}\left[\frac{X_{A\cap B}}{X_{A\cap B} + X_{A\setminus B}} \unicode{x1D7D9}_{X_{A\cap B} + X_{A\setminus B}>0},\right] &= \mathrm{E}\left[\int_0^1 X_{A\cap B} \,  t^{X_{A\cap B}+X_{A\setminus B}-1}\,dt\right]\\
    &= \int_0^1 \mathrm{E}\left[X_{A\cap B}  \, t^{X_{A\cap B}+X_{A\setminus B}-1}\,dt\right]\label{line:1}\\
    &= \int_0^1 \mathrm{E}\left[X_{A\cap B}  \, t^{X_{A\cap B}-1}\right] \mathrm{E}\left[t^X_{A\setminus B}\right]\,dt\label{line:2}\\
    &= \vert A\cap B\vert \int_0^1(1-s+st)^{\vert A\cap B \vert + \vert A\setminus B \vert -1}\, dt\\
    &= \frac{\vert A \cap B\vert (1-s+st)^{\vert A \vert}}{\vert A \vert}\bigg\rvert_{t=0}^{t=1}\\
    &= \frac{\vert A\cap B \vert}{\vert A \vert} \left(1-(1-s)^{\vert A\vert}\right),
\end{align}
```
{% endraw %}

using Fubini's theorem and independence.


In light of Theorem 1, we note that $\hat{C}_\text{frac}(A,B)$ is *not* an unbiased
estimate of $C(A,B)$. This may explain the observations in
[@luiz_thesis] that show suboptimal performance for
short sequences (e.g. viruses). However, for sufficiently large $\vert
A \vert$ and $s$, the bias factor $\left(1-(1-s)^{\vert
A\vert}\right)$ is sufficiently close to 1.

Hence we can define:

$$
C_\text{frac}(A,B)
= \frac{\vert A\cap B \vert}{\vert A \vert} \left(1-(1-s)^{\vert A\vert}\right)^{-1}
$$
which will have expectation

$$
\mathrm{E} [C_\text{frac}(A,B)] = \frac{\vert A\cap B \vert}{\vert A \vert}
$$
by Theorem 1.

## Implementation of FracMinHash and min-set-cov

We provide implementations of FracMinHash and min-set-cov in the
software package `sourmash`, which is implemented in Python and Rust
and developed under the BSD license
[@sourmash_joss]. FracMinHash sketches were created for DNA
sequence inputs using the `sourmash sketch dna` command with the
`scaled` parameter. Minimum metagenome covers were generated using
`sourmash gather` with the sketched metagenome as query against a
collection of one or more sketched genomes.

sourmash is available at
[github.com/sourmash-bio/sourmash](https://github.com/sourmash-bio/sourmash/). The
results in this paper were generated with sourmash v4.2.3.

<!-- CTB: put DOI in when zenodo is fixed -->

## Comparison between CMash, mash screen, and Scaled MinHash.

Experiments use $k=\{21, 31, 51\}$ (except for Mash, which only
supports $k \le 32$).  For Mash and CMash they were run with
$n=\{1000, 10000\}$ to evaluate the containment estimates when using
larger sketches with sizes comparable to the FracMinHash sketches
with $scaled=1000$.  The truth set is calculated using an exact
$k$-mer counter implemented with a _HashSet_ data structure in the
Rust programming language [@doi:10.1145/2692956.2663188].
The sourmash results were generated with `sourmash search --containment`.

For _Mash Screen_ the ratio of hashes matched by total hashes is used
instead of the _Containment Score_, since the latter uses a $k$-mer
survival process modeled as a Poisson process first introduced in
[@doi:10.1186/s12864-015-1647-5] and later used in the _Mash distance_
[@doi:10.1186/s13059-016-0997-x] 
and _Containment score_ [@doi:10.1186/s13059-019-1841-x]
formulations.

<!-- CTB: include performance information. -->

## GenBank database sketching and searches

Minimum metagenome covers were calculated using a microbial genome
subset of GenBank (July 2020, 725,339 genomes) using a scaled
factor of 2000 and a k-mer size of 31. Sketches for all genomes and
metagenomes were calculated with `sourmash sketch dna -p
scaled=2000,k=31`. The minimum metagenome covers were calculated using
all genomes sharing 50 hashes with the metagenome (that is, an
estimated overlap of 100,000 k-mers) with
`sourmash gather --threshold-bp 1e5`.
Overlapping sketches were saved with `--save-prefetch`
and matches were saved with `--save-matches`.

The GenBank database used is 24 GB in size and is available for download
through the sourmash project [@url:https://sourmash.bio].

<!-- CTB: include performance information. -->

## Taxonomy

The CAMI evaluations were run with the sourmash CAMI pipeline
[@url:https://github.com/luizirber/2020-cami] against
the Jan 8, 2019 RefSeq snapshot provided by CAMI.  This pipeline generated
Open-community Profiling Assessment (OPAL) compatible output
[@doi:10.1038/nmeth.4458].
This output was then processed with the standard CAMI tools.

## Read mapping and hybrid mapping pipeline

Metagenome reads were mapped to reference genomes using minimap2 v2.17
[@doi:10.1093/bioinformatics/bty191] with short single-end read mapping mode
(`-x sr`).

The hybrid selection and mapping pipeline using the rank-ordered min-set-cov
results was implemented in the `subtract_gather.py` script
in the genome-grist package [@url:https://github.com/dib-lab/genome-grist].

The complete workflow, from metagenome download to taxonomic analysis
and iterative mapping, is implemented in the genome-grist package.
genome-grist uses snakemake [@doi:10.12688/f1000research.29032.2] to
define and execute a workflow that combines sourmash sketching,
metagenome cover calculation, and taxonomic analysis with metagenome
download from the SRA, genome download from GenBank, and read mapping.
We used genome-grist v0.7.4
[@url:https://dx.doi.org/10.5281/zenodo.5792144] to generate the
results in this paper; see `conf-paper.yml` in the pipeline
repository.

genome-grist relies on matplotlib [@doi:10.1109/MCSE.2007.55], Jupyter
Notebook [@url:https://dx.doi.org/10.3233/978-1-61499-649-1-87], numpy
[@doi:10.1038/s41586-020-2649-2], pandas
[@url:https://dx.doi.org/10.5281/zenodo.3509134], papermill, samtools
[@doi:10.1093/bioinformatics/btp352], bedtools
[@doi:10.1093/bioinformatics/btq033], fastp
[@doi:10.1093/bioinformatics/bty560], khmer
[@doi:10.21105/joss.00272], screed
[@url:https://screed.readthedocs.io], seqtk
[@url:https://github.com/lh3/seqtk], and sra-tools
[@url:https://github.com/ncbi/sra-tools].  These tools are all
installed and managed in snakemake via conda
[@url:https://docs.anaconda.com] and bioconda
[@doi:10.1038/s41592-018-0046-7].  genome-grist itself is developed under
the BSD 3-clause open source license, and is available at
[github.com/dib-lab/genome-grist/](https://github.com/dib-lab/genome-grist/).

## Intermediate data products and figure generation

All figures were generated using the Jupyter Notebooks from v0.1 of
the
[github.com/dib-lab/2021-paper-sourmash-gather-pipeline](github.com/dib-lab/2021-paper-sourmash-gather-pipeline)
repository [@url:https://dx.doi.org/10.5281/zenodo.5793387]. This
repository also contains the intermediate data products necessary for
figure generation.

## Metagenome data set accessions

The accessions for the metagenome data sets in Table
@tbl:genbank-cover are:

| data set | SRA accession |
| -------- | ------------- |
| `zymo mock` | SRR12324253 |
| `podar mock` | SRR606249 |
| `gut real` | SRR5650070  |
| `oil well real`  | SRR1976948 |
